File: src/main/java/com/imbananko/tilly/repository/VoteRepository.java
package com.imbananko.tilly.repository;

import com.imbananko.tilly.model.VoteEntity;
import com.imbananko.tilly.model.VoteEntity.Value;
import com.imbananko.tilly.utility.SqlQueries;
import io.vavr.Tuple2;
import io.vavr.collection.HashMap;
import io.vavr.collection.Map;
import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
import org.springframework.stereotype.Repository;

@Repository
public class VoteRepository {

  private final NamedParameterJdbcTemplate template;
  private final Map<String, String> queries;

  public VoteRepository(NamedParameterJdbcTemplate template, SqlQueries queries) {
    this.template = template;
    this.queries = queries.getQueries();
  }

  @SuppressWarnings("ConstantConditions")
  public boolean exists(VoteEntity vote) {
    return template.queryForObject(queries.getOrElse("voteExists", null), getParams(vote), Boolean.class);
  }

  public void insertOrUpdate(VoteEntity vote) {
    template.update(queries.getOrElse("insertOrUpdateVote", null), getParams(vote));
  }

  public void delete(VoteEntity vote) {
    template.update(queries.getOrElse("deleteVote", null), getParams(vote));
  }

  public HashMap<Value, Long> getStats(long chatId, int messageId) {
    return HashMap.ofEntries(template.query(queries.getOrElse("findVoteStats", null),
      new MapSqlParameterSource("chatId", chatId).addValue("messageId", messageId),
      (rs, rowNum) ->
        new Tuple2<>(Value.valueOf(rs.getString("value")), rs.getLong("count"))));
  }

  private MapSqlParameterSource getParams(VoteEntity vote) {
    return new MapSqlParameterSource("chatId", vote.getChatId())
      .addValue("messageId", vote.getMessageId())
      .addValue("voterId", vote.getVoterId())
      .addValue("value", vote.getValue().name());
  }
}

File: src/main/kotlin/com/imbananko/tilly/repository/VoteRepository.kt
package com.imbananko.tilly.repository

import com.imbananko.tilly.model.MemeEntity
import com.imbananko.tilly.model.MemeStatsEntry
import com.imbananko.tilly.model.VoteEntity
import com.imbananko.tilly.model.VoteValue.*
import com.imbananko.tilly.utility.SqlQueries
import org.springframework.jdbc.core.namedparam.MapSqlParameterSource
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import org.springframework.stereotype.Repository

@Repository
class VoteRepository(private val template: NamedParameterJdbcTemplate, private val queries: SqlQueries) {
  fun insertOrUpdate(vote: VoteEntity): Int = template.update(queries.getFromConfOrFail("insertOrUpdateVote"), getParams(vote))

  fun delete(vote: VoteEntity) = template.update(queries.getFromConfOrFail("deleteVote"), getParams(vote))

  fun getStatsByUser(chatId: Long, userId: Int): List<MemeStatsEntry> = template.query(
      queries.getFromConfOrFail("findUserStats"),
      MapSqlParameterSource("chatId", chatId).addValue("userId", userId))
  { rs, _ ->
    MemeStatsEntry(
        UP to rs.getInt(UP.name),
        DOWN to rs.getInt(DOWN.name))
  }.toList()

  fun getVotes(meme: MemeEntity): List<VoteEntity> = template.query(
      queries.getFromConfOrFail("findMemeVotes"),
      MapSqlParameterSource("chatId", meme.chatId).addValue("messageId", meme.messageId))
  { rs, _ ->
    VoteEntity(
        rs.getLong("chat_id"),
        rs.getInt("message_id"),
        rs.getInt("voter_id"),
        valueOf(rs.getString("value")))
  }.toList()

  private fun getParams(vote: VoteEntity): MapSqlParameterSource =
      MapSqlParameterSource("chatId", vote.chatId)
          .addValue("messageId", vote.messageId)
          .addValue("voterId", vote.voterId)
          .addValue("value", vote.voteValue.name)
}
